### PR TITLE
Update DJS + Display names fixes + per-server avatars

### PR DESCRIPTION
**Pull request by metal0**
Fixes ZDEV-39
Fixes ZDEV-57

Supersedes #416

### CODE REVIEW

File: backend/src/plugins/Automod/triggers/roleRemoved.ts [hunk 1]
```diff
@@ -38,7 +38,7 @@ export const RoleRemovedTrigger = automodTrigger<RoleAddedMatchResult>()({
     const role = pluginData.guild.roles.cache.get(matchResult.extra.matchedRoleId as Snowflake);
     const roleName = role?.name || "Unknown";
     const member = contexts[0].member!;
-    const memberName = `**${renderUserUsername(member.user)}** (\`${member.id}\`)`;
+    const memberName = `**${renderUsername(member.user)}** (\`${member.id}\`)`;
```
Comment (almeidx) [inline backend/src/plugins/Automod/triggers/roleRemoved.ts:41]:
```suggestion
    const memberName = `**${renderUsername(member)}** (\`${member.id}\`)`;
```

File: backend/src/plugins/BotControl/commands/ServersCmd.ts [hunk 2]
```diff
@@ -48,7 +48,10 @@ export const ServersCmd = botControlCmd({
         const lines = filteredGuilds.map((g) => {
           const paddedId = g.id.padEnd(longestId, " ");
           const owner = getUser(pluginData.client, g.ownerId);
-          return `\`${paddedId}\` **${g.name}** (${g.memberCount} members) (owner **${owner.tag}** \`${owner.id}\`)`;
+          return `\`${paddedId}\` **${g.name}** (${g.memberCount} members) (owner **${renderUsername(
+            owner.username,
+            owner.discriminator,
+          )}** \`${owner.id}\`)`;
```
Comment (almeidx) [inline backend/src/plugins/BotControl/commands/ServersCmd.ts:54]:
```suggestion
          return `\`${paddedId}\` **${g.name}** (${g.memberCount} members) (owner **${renderUsername(owner)}** \`${owner.id}\`)`;
```

File: backend/src/plugins/Cases/functions/createCase.ts [hunk 3]
```diff
@@ -1,23 +1,23 @@
...
-import { renderUserUsername, resolveUser } from "../../../utils";
+import { renderUsername, resolveUser } from "../../../utils";
 import { CaseArgs, CasesPluginType } from "../types";
 import { createCaseNote } from "./createCaseNote";
 import { postCaseToCaseLogChannel } from "./postToCaseLogChannel";
 
 export async function createCase(pluginData: GuildPluginData<CasesPluginType>, args: CaseArgs) {
   const user = await resolveUser(pluginData.client, args.userId);
-  const userName = renderUserUsername(user);
+  const userName = renderUsername(user.username, user.discriminator);
```
Comment (almeidx) [inline backend/src/plugins/Cases/functions/createCase.ts:11]: "userName" :cringe:
```suggestion
  const userName = renderUsername(user);
```

File: backend/src/plugins/Cases/functions/createCase.ts [hunk 4]
```diff
@@ -1,23 +1,23 @@
...
 
   const mod = await resolveUser(pluginData.client, args.modId);
-  const modName = mod.tag;
+  const modName = renderUsername(mod.username, mod.discriminator);
```
Comment (almeidx) [inline backend/src/plugins/Cases/functions/createCase.ts:14]:
```suggestion
  const modName = renderUsername(mod);
```

File: backend/src/plugins/Cases/functions/createCase.ts [hunk 5]
```diff
@@ -1,23 +1,23 @@
...
 
   let ppName: string | null = null;
   let ppId: Snowflake | null = null;
   if (args.ppId) {
     const pp = await resolveUser(pluginData.client, args.ppId);
-    ppName = pp.tag;
+    ppName = renderUsername(pp.username, pp.discriminator);
```
Comment (almeidx) [inline backend/src/plugins/Cases/functions/createCase.ts:20]:
```suggestion
    ppName = renderUsername(pp);
```

File: backend/src/plugins/ModActions/commands/CasesModCmd.ts [hunk 6]
```diff
@@ -28,8 +28,10 @@ export const CasesModCmd = modActionsCmd({
 
   async run({ pluginData, message: msg, args }) {
     const modId = args.mod || msg.author.id;
-    const mod = await resolveUser(pluginData.client, modId);
-    const modName = mod instanceof User ? mod.tag : modId;
+    const mod =
+      (await resolveMember(pluginData.client, pluginData.guild, modId)) ||
+      (await resolveUser(pluginData.client, modId));
+    const modName = mod instanceof User ? renderUsername(mod) : modId;
```
Comment (almeidx) [inline backend/src/plugins/ModActions/commands/CasesModCmd.ts:34]: This doesn't account for GuildMember, maybe check the opposite instead (you have to import it)
```suggestion
    const modName = mod instanceof UnknownUser ? modId : renderUsername(mod);
```

File: backend/src/plugins/ModActions/commands/CasesModCmd.ts [hunk 7]
```diff
@@ -57,7 +59,7 @@ export const CasesModCmd = modActionsCmd({
         const embed = {
           author: {
             name: title,
-            icon_url: mod instanceof User ? mod.displayAvatarURL() : undefined,
+            icon_url: mod instanceof User || mod instanceof GuildMember ? mod.displayAvatarURL() : undefined,
```
Comment (almeidx) [inline backend/src/plugins/ModActions/commands/CasesModCmd.ts:62]:
```suggestion
            icon_url: mod instanceof UnknownUser ? undefined : mod.displayAvatarURL(),
```

File: backend/src/plugins/ModActions/commands/CasesModCmd.ts [hunk 8]
```diff
@@ -1,7 +1,7 @@
-import { APIEmbed, User } from "discord.js";
+import { APIEmbed, GuildMember, User } from "discord.js";
```
Comment (almeidx) [inline backend/src/plugins/ModActions/commands/CasesModCmd.ts:1]:
```suggestion
import { APIEmbed, User } from "discord.js";
```

File: backend/src/plugins/ModActions/commands/CasesUserCmd.ts [hunk 9]
```diff
@@ -35,8 +43,10 @@ export const CasesUserCmd = modActionsCmd({
   ],
 
   async run({ pluginData, message: msg, args }) {
-    const user = await resolveUser(pluginData.client, args.user);
-    if (!user.id) {
+    const user =
+      (await resolveMember(pluginData.client, pluginData.guild, args.user)) ||
+      (await resolveUser(pluginData.client, args.user));
+    if (!user.id || user instanceof UnknownUser) {
```
Comment (almeidx) [inline backend/src/plugins/ModActions/commands/CasesUserCmd.ts:49]: Should be enough
```suggestion
    if (user instanceof UnknownUser) {
```

File: backend/src/plugins/ModActions/events/PostAlertOnMemberJoinEvt.ts [hunk 10]
```diff
@@ -46,7 +46,7 @@ export const PostAlertOnMemberJoinEvt = modActionsEvt({
       }
 
       await alertChannel.send(
-        `<@!${member.id}> (${renderUserUsername(member.user)} \`${member.id}\`) joined with ${
+        `<@!${member.id}> (${renderUsername(member.user)} \`${member.id}\`) joined with ${
```
Comment (almeidx) [inline backend/src/plugins/ModActions/events/PostAlertOnMemberJoinEvt.ts:49]:
```suggestion
        `<@!${member.id}> (${renderUsername(member)} \`${member.id}\`) joined with ${
```

File: backend/src/plugins/ModActions/commands/WarnCmd.ts [hunk 11]
```diff
@@ -106,7 +106,7 @@ export const WarnCmd = modActionsCmd({
     sendSuccessMessage(
       pluginData,
       msg.channel,
-      `Warned **${renderUserUsername(memberToWarn.user)}** (Case #${warnResult.case.case_number})${messageResultText}`,
+      `Warned **${renderUsername(memberToWarn.user)}** (Case #${warnResult.case.case_number})${messageResultText}`,
```
Comment (almeidx) [inline backend/src/plugins/ModActions/commands/WarnCmd.ts:109]:
```suggestion
      `Warned **${renderUsername(memberToWarn)}** (Case #${warnResult.case.case_number})${messageResultText}`,
```

File: backend/src/plugins/ModActions/functions/actualKickMemberCmd.ts [hunk 12]
```diff
@@ -103,7 +103,7 @@ export async function actualKickMemberCmd(
   }
 
   // Confirm the action to the moderator
-  let response = `Kicked **${renderUserUsername(memberToKick.user)}** (Case #${kickResult.case.case_number})`;
+  let response = `Kicked **${renderUsername(memberToKick.user)}** (Case #${kickResult.case.case_number})`;
```
Comment (almeidx) [inline backend/src/plugins/ModActions/functions/actualKickMemberCmd.ts:106]:
```suggestion
  let response = `Kicked **${renderUsername(memberToKick)}** (Case #${kickResult.case.case_number})`;
```

File: backend/src/plugins/Mutes/commands/MutesCmd.ts [hunk 13]
```diff
@@ -74,7 +74,7 @@ export const MutesCmd = mutesCmd({
       totalMutes = manuallyMutedMembers.length;
 
       lines = manuallyMutedMembers.map((member) => {
-        return `<@!${member.id}> (**${renderUserUsername(member.user)}**, \`${member.id}\`)   🔧 Manual mute`;
+        return `<@!${member.id}> (**${renderUsername(member.user)}**, \`${member.id}\`)   🔧 Manual mute`;
```
Comment (almeidx) [inline backend/src/plugins/Mutes/commands/MutesCmd.ts:77]:
```suggestion
        return `<@!${member.id}> (**${renderUsername(member)}**, \`${member.id}\`)   🔧 Manual mute`;
```

File: backend/src/plugins/Mutes/commands/MutesCmd.ts [hunk 14]
```diff
@@ -123,7 +123,7 @@ export const MutesCmd = mutesCmd({
 
       lines = filteredMutes.map((mute) => {
         const user = pluginData.client.users.resolve(mute.user_id as Snowflake);
-        const username = user ? renderUserUsername(user) : "Unknown#0000";
+        const username = user ? renderUsername(user) : "Unknown#0000";
```
Comment (almeidx) [inline backend/src/plugins/Mutes/commands/MutesCmd.ts:126]:
```suggestion
        const username = renderUsername(user);
```

File: backend/src/plugins/ReactionRoles/util/addMemberPendingRoleChange.ts [hunk 15]
```diff
@@ -33,9 +33,7 @@ export async function addMemberPendingRoleChange(
           try {
             await member.roles.set(Array.from(newRoleIds.values()), "Reaction roles");
           } catch (e) {
-            logger.warn(
-              `Failed to apply role changes to ${renderUserUsername(member.user)} (${member.id}): ${e.message}`,
-            );
+            logger.warn(`Failed to apply role changes to ${renderUsername(member.user)} (${member.id}): ${e.message}`);
```
Comment (almeidx) [inline backend/src/plugins/ReactionRoles/util/addMemberPendingRoleChange.ts:36]: Pretty sure there is a "renderUserInfo()" function somewhere that renders "${tag} ($i{d})"
```suggestion
            logger.warn(`Failed to apply role changes to ${renderUsername(member)} (${member.id}): ${e.message}`);
```

File: backend/src/plugins/Starboard/util/createStarboardEmbedFromMessage.ts [hunk 16]
```diff
@@ -28,7 +28,7 @@ export function createStarboardEmbedFromMessage(
     embed.color = color;
   }
 
-  embed.author.icon_url = msg.author.displayAvatarURL();
+  embed.author.icon_url = (msg.member || msg.author).displayAvatarURL();
```
Comment (almeidx) [inline backend/src/plugins/Starboard/util/createStarboardEmbedFromMessage.ts:31]:
```suggestion
  embed.author.icon_url = (msg.member ?? msg.author).displayAvatarURL();
```

File: backend/src/plugins/Utility/commands/AvatarCmd.ts [hunk 17]
```diff
@@ -10,17 +10,17 @@ export const AvatarCmd = utilityCmd({
   permission: "can_avatar",
 
   signature: {
-    user: ct.resolvedUserLoose({ required: false }),
+    user: ct.resolvedMember({ required: false }) || ct.resolvedUserLoose({ required: false }),
   },
 
   async run({ message: msg, args, pluginData }) {
-    const user = args.user || msg.author;
+    const user = args.user || msg.member || msg.author;
```
Comment (almeidx) [inline backend/src/plugins/Utility/commands/AvatarCmd.ts:17]:
```suggestion
    const user = args.user ?? msg.member ?? msg.author;
```

File: backend/src/plugins/Utility/commands/LevelCmd.ts [hunk 18]
```diff
@@ -18,6 +18,6 @@ export const LevelCmd = utilityCmd({
   run({ message, args, pluginData }) {
     const member = args.member || message.member;
     const level = getMemberLevel(pluginData, member);
-    message.channel.send(`The permission level of ${renderUserUsername(member.user)} is **${level}**`);
+    message.channel.send(`The permission level of ${renderUsername(member.user)} is **${level}**`);
```
Comment (almeidx) [inline backend/src/plugins/Utility/commands/LevelCmd.ts:21]:
```suggestion
    message.channel.send(`The permission level of ${renderUsername(member)} is **${level}**`);
```

File: backend/src/plugins/Utility/commands/VcdisconnectCmd.ts [hunk 19]
```diff
@@ -43,7 +43,7 @@ export const VcdisconnectCmd = utilityCmd({
     sendSuccessMessage(
       pluginData,
       msg.channel,
-      `**${renderUserUsername(args.member.user)}** disconnected from **${channel.name}**`,
+      `**${renderUsername(args.member.user)}** disconnected from **${channel.name}**`,
```
Comment (almeidx) [inline backend/src/plugins/Utility/commands/VcdisconnectCmd.ts:46]:
```suggestion
      `**${renderUsername(args.member)}** disconnected from **${channel.name}**`,
```

File: backend/src/plugins/Utility/commands/VcmoveCmd.ts [hunk 20]
```diff
@@ -80,11 +80,7 @@ export const VcmoveCmd = utilityCmd({
       newChannel: channel,
     });
 
-    sendSuccessMessage(
-      pluginData,
-      msg.channel,
-      `**${renderUserUsername(args.member.user)}** moved to **${channel.name}**`,
-    );
+    sendSuccessMessage(pluginData, msg.channel, `**${renderUsername(args.member.user)}** moved to **${channel.name}**`);
```
Comment (almeidx) [inline backend/src/plugins/Utility/commands/VcmoveCmd.ts:83]:
```suggestion
    sendSuccessMessage(pluginData, msg.channel, `**${renderUsername(args.member)}** moved to **${channel.name}**`);
```

File: backend/src/plugins/Utility/commands/VcmoveCmd.ts [hunk 21]
```diff
@@ -157,7 +153,7 @@ export const VcmoveAllCmd = utilityCmd({
         sendErrorMessage(
           pluginData,
           msg.channel,
-          `Failed to move ${renderUserUsername(currMember.user)} (${currMember.id}): You cannot act on this member`,
+          `Failed to move ${renderUsername(currMember.user)} (${currMember.id}): You cannot act on this member`,
```
Comment (almeidx) [inline backend/src/plugins/Utility/commands/VcmoveCmd.ts:156]:
```suggestion
          `Failed to move ${renderUsername(currMember)} (${currMember.id}): You cannot act on this member`,
```

File: backend/src/plugins/Utility/commands/VcmoveCmd.ts [hunk 22]
```diff
@@ -175,7 +171,7 @@ export const VcmoveAllCmd = utilityCmd({
         sendErrorMessage(
           pluginData,
           msg.channel,
-          `Failed to move ${renderUserUsername(currMember.user)} (${currMember.id})`,
+          `Failed to move ${renderUsername(currMember.user)} (${currMember.id})`,
```
Comment (almeidx) [inline backend/src/plugins/Utility/commands/VcmoveCmd.ts:174]:
```suggestion
          `Failed to move ${renderUsername(currMember)} (${currMember.id})`,
```

File: backend/src/plugins/Utility/functions/getUserInfoEmbed.ts [hunk 23]
```diff
@@ -40,13 +39,11 @@ export async function getUserInfoEmbed(
...
 
-  const timeAndDate = pluginData.getPlugin(TimeAndDatePlugin);
-
   embed.author = {
-    name: `${user.bot ? "Bot" : "User"}:  ${renderUsername(user.username, user.discriminator)}`,
+    name: `${user.bot ? "Bot" : "User"}:  ${renderUsername(user)}`,
   };
 
-  const avatarURL = user.displayAvatarURL();
+  const avatarURL = (member || user).displayAvatarURL();
```
Comment (almeidx) [inline backend/src/plugins/Utility/functions/getUserInfoEmbed.ts:46]:
```suggestion
  const avatarURL = (member ?? user).displayAvatarURL();
```

File: backend/src/plugins/Utility/search.ts [hunk 24]
```diff
@@ -381,7 +381,7 @@ async function performMemberSearch(
         return true;
       }
 
-      const fullUsername = renderUserUsername(member.user);
+      const fullUsername = renderUsername(member.user);
```
Comment (almeidx) [inline backend/src/plugins/Utility/search.ts:384]:
```suggestion
      const fullUsername = renderUsername(member);
```

File: backend/src/plugins/Utility/search.ts [hunk 25]
```diff
@@ -492,10 +492,10 @@ function formatSearchResultList(members: Array<GuildMember | User>): string {
     const paddedId = member.id.padEnd(longestId, " ");
     let line;
     if (member instanceof GuildMember) {
-      line = `${paddedId} ${renderUserUsername(member.user)}`;
+      line = `${paddedId} ${renderUsername(member.user)}`;
```
Comment (almeidx) [inline backend/src/plugins/Utility/search.ts:495]:
```suggestion
      line = `${paddedId} ${renderUsername(member)}`;
```

**Review by almeidx — CHANGES_REQUESTED** [review]

**Review by almeidx — APPROVED** [review]

**Comment by Dragory** [comment]
Continued in #453

**Pull request closed without merging**